src/contexts/UserContext.tsx

```diff
@@ -0,0 +1,40 @@
+'use client'
+
+import { onAuthStateChangedListener } from '@/api/firebase'
+import { User, UserCredential } from 'firebase/auth'
+import { createContext, useState, useEffect, ReactNode, Dispatch, SetStateAction } from 'react'
+
+type UserContextType = {
+    currentUser: User | null
+    setCurrentUser: Dispatch<SetStateAction<User | null>>
+}
+
+type Props = {
+    children: ReactNode
+}
+
+export const UserContext = createContext<UserContextType>({
+    setCurrentUser: () => null,
+    currentUser: null
+})
+
+export const UserProvider = ({ children }: Props) => {
+    const [currentUser, setCurrentUser] = useState<User | null>(null)
+    const value = {
+        currentUser,
+        setCurrentUser
+    }
+
+    useEffect(() => {
+        const unsubscribe = onAuthStateChangedListener((user) => {
```

Comment:
`unsubscribe` is always undefined? _onAuthStateChangedListener()_ has no return statement.

Reply:
Ah yes, good call. onAuthStateChangedListener() returns the firebase.Unsubscribe interface that is returned by firebase's onAuthStateChanged method. I added this type to onAuthStateChangedListener, as well as a return, so the unsubscribe inside of the useEffect is now also typed.